raiden/raiden_event_handler.py

```diff
@@ -341,56 +370,135 @@ def handle_contract_send_channelunlock(
         payment_channel: PaymentChannel = raiden.chain.payment_channel(
             canonical_identifier=canonical_identifier,
         )
-        token_network: TokenNetwork = payment_channel.token_network
 
-        participants_details = token_network.detail_participants(
-            participant1=raiden.address,
-            participant2=participant,
-            block_identifier=triggered_by_block_hash,
-            channel_identifier=canonical_identifier.channel_identifier,
+        channel_state = get_channelstate_by_token_network_and_partner(
+            chain_state=state_from_raiden(raiden),
+            token_network_id=token_network_identifier,
+            partner_address=participant,
         )
 
-        our_details = participants_details.our_details
-        our_locksroot = our_details.locksroot
+        if not channel_state:
+            # channel was cleaned up already due to an unlock
+            return
 
-        partner_details = participants_details.partner_details
-        partner_locksroot = partner_details.locksroot
+        our_address = channel_state.our_state.address
+        our_locksroot = channel_state.our_state.onchain_locksroot
+
+        partner_address = channel_state.partner_state.address
+        partner_locksroot = channel_state.partner_state.onchain_locksroot
+
+        # During a migration, a channel might have been settled
+        # but some funds might still be unlocked. At migration
+        # time, it's pretty hard to find the partner's address
+        # in a reliable way so it was opted here that the values
+        # are populated by None values where we re-query the contract
+        # here to be able to properly unlock funds.
+        if our_locksroot is None or partner_locksroot is None:
```

Comment:
Okay, so this is a bit annoying of me, but if we can't have these values in the `ContractSendBatchUnlock`, why don't we have this some sort of callback to the `ContractReceiveChannelSettled`? The migration seems quite pointless without it, no?

Reply:
To clarify, @rakanalh found a corner case that requires the blockchain query.

We must query the `locksroot` from the blockchain on the channel settled event, the reason being that we don't know if our latest message was received by the target or not, and this can happen:

- A sends a Locked transfer to B
- The lock expires
- B goes offline
- A sends to B the RemovedExpiredLock (From A's perspective we don't have any pending locks)
- The channel is closed by A
- B's monitoring service uses the Locked transfer balance proof (The blockchain state is not the same as our local state)

From A's perspective, the channel does not have any pending locks in the merkle tree, which is not true, so to avoid wrongly clearing the channel from the local state the locksroot is necessary.